File: modules/agent/src/schemas/task.py
"""Task schemas for API validation."""

from datetime import date
from decimal import Decimal
from typing import Optional

from pydantic import BaseModel, field_validator


FIBONACCI_VALUES = (1, 2, 3, 5, 8, 13, 21)


class TaskCreate(BaseModel):
    title: str
    description: Optional[str] = None
    taskable_type: Optional[str] = None
    taskable_id: Optional[int] = None
    current_status_id: Optional[int] = None
    priority_id: Optional[int] = None
    start_date: Optional[date] = None
    due_date: Optional[date] = None
    estimated_hours: Optional[Decimal] = None
    actual_hours: Optional[Decimal] = None
    complexity: Optional[int] = None
    sort_order: Optional[int] = None
    assigned_to_individual_id: Optional[int] = None

    @field_validator("complexity")
    @classmethod
    def validate_complexity(cls, v):
        if v is not None and v not in FIBONACCI_VALUES:
            raise ValueError(f"complexity must be one of {FIBONACCI_VALUES}")
        return v


class TaskUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    taskable_type: Optional[str] = None
    taskable_id: Optional[int] = None
    current_status_id: Optional[int] = None
    priority_id: Optional[int] = None
    start_date: Optional[date] = None
    due_date: Optional[date] = None
    estimated_hours: Optional[Decimal] = None
    actual_hours: Optional[Decimal] = None
    complexity: Optional[int] = None
    sort_order: Optional[int] = None
    completed_at: Optional[str] = None
    assigned_to_individual_id: Optional[int] = None

    @field_validator("complexity")
    @classmethod
    def validate_complexity(cls, v):
        if v is not None and v not in FIBONACCI_VALUES:
            raise ValueError(f"complexity must be one of {FIBONACCI_VALUES}")
        return v

File: modules/agent/internal/schemas/task.go
package schemas

// TaskCreate represents the request body for creating a task
type TaskCreate struct {
	Title                  string   `json:"title" validate:"required"`
	Description            *string  `json:"description"`
	TaskableType           *string  `json:"taskable_type" validate:"omitempty,oneof=Deal Lead Job Project Organization Individual Contact"`
	TaskableID             *int64   `json:"taskable_id"`
	CurrentStatusID        *int64   `json:"current_status_id"`
	PriorityID             *int64   `json:"priority_id"`
	StartDate              *string  `json:"start_date"`
	DueDate                *string  `json:"due_date"`
	EstimatedHours         *float64 `json:"estimated_hours" validate:"omitempty,min=0"`
	ActualHours            *float64 `json:"actual_hours" validate:"omitempty,min=0"`
	Complexity             *int16   `json:"complexity" validate:"omitempty,oneof=1 2 3 5 8 13 21"`
	SortOrder              *int     `json:"sort_order"`
	AssignedToIndividualID *int64   `json:"assigned_to_individual_id"`
}

// TaskUpdate represents the request body for updating a task
type TaskUpdate struct {
	Title                  *string  `json:"title"`
	Description            *string  `json:"description"`
	TaskableType           *string  `json:"taskable_type" validate:"omitempty,oneof=Deal Lead Job Project Organization Individual Contact"`
	TaskableID             *int64   `json:"taskable_id"`
	CurrentStatusID        *int64   `json:"current_status_id"`
	PriorityID             *int64   `json:"priority_id"`
	StartDate              *string  `json:"start_date"`
	DueDate                *string  `json:"due_date"`
	EstimatedHours         *float64 `json:"estimated_hours" validate:"omitempty,min=0"`
	ActualHours            *float64 `json:"actual_hours" validate:"omitempty,min=0"`
	Complexity             *int16   `json:"complexity" validate:"omitempty,oneof=1 2 3 5 8 13 21"`
	SortOrder              *int     `json:"sort_order"`
	CompletedAt            *string  `json:"completed_at"`
	AssignedToIndividualID *int64   `json:"assigned_to_individual_id"`
}
